rust/src/mqtt/mqtt.rs

```diff
@@ -0,0 +1,846 @@
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingPublish);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::PUBACK(ref v)
+            | MQTTOperation::PUBCOMP(ref v) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                if let Some(tx) = self.get_tx_by_pkt_id(v.message_id as u32) {
+                    (*tx).msg.push(msg);
+                    (*tx).complete = true;
+                    (*tx).pkt_id = None;
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingPublish);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::SUBACK(ref suback) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                if let Some(tx) = self.get_tx_by_pkt_id(suback.message_id as u32) {
+                    (*tx).msg.push(msg);
+                    (*tx).complete = true;
+                    (*tx).pkt_id = None;
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingSubscribe);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::UNSUBACK(ref unsuback) => {
+                if !self.connected {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::UnintroducedMessage);
+                    self.transactions.push(tx);
+                    return;
+                }
+                if let Some(tx) = self.get_tx_by_pkt_id(unsuback.message_id as u32) {
+                    (*tx).msg.push(msg);
+                    (*tx).complete = true;
+                    (*tx).pkt_id = None;
+                } else {
+                    let mut tx = self.new_tx(msg, toclient);
+                    &mut MQTTState::set_event(&mut tx, MQTTEvent::MissingUnsubscribe);
+                    self.transactions.push(tx);
+                }
+            },
+            MQTTOperation::UNASSIGNED => {
+                let mut tx = self.new_tx(msg, toclient);
+                tx.complete = true;
+                &mut MQTTState::set_event(&mut tx, MQTTEvent::UnassignedMsgtype);
+                self.transactions.push(tx);
+            },
+            MQTTOperation::TRUNCATED(_) => {
```

Comment:
I am wondering if there is more to do for certain messages type...

I think we can expect certain messages type to be truncated like `PUBLISH`, but we do not expect a truncated `CONNECT` 

What are your thoughts on this ?

Reply:
I think we do. It is certainly possible to have a `CONNECT` with lots of big user properties in MQTT5, which could also easily hit the limit. I think we should always only consider the length specified in the header and not make assumptions based on the message type.
All types have the same problem though: to get the packet identifier (or message ID) that allows us to tie transactions together, we need to parse the message. The ID comes after the properties which by themselves could be huge already.

Reply:
Thanks for the insights :-)